### ISSUE TITLE
Auto Indirect Reference Inference

### ISSUE THREAD

**Issue by cooperjbrandon** [infinum/datx#199]
Submitting a:

* [ ] Bug report
* [ ] Feature suggestion
* [x] Support request

--------------

Is there a way to auto support indirect references (or does datx already support this)? I might be using the wrong lexicon here. This is what I mean:

Simple example of a User model and Post model:

```
class Post extends jsonapi(Model) {
  public static type = 'post';

  @prop.identifier public id!: number;
  @prop public title!: string;

  @prop.toOne('user') public author!: User;
}

class User extends jsonapi(Model) {
  public static type = 'user';

  @prop.identifier public id!: number;
  @prop public email!: string;

  @prop.toMany('post') public posts!: Post;
}
```
Then when doing:
```
const user33  = collection.add({id: 33, email: 'a@b.com', posts:[22]}, 'user');
const post22  = collection.add({id: 22, title: 'my_title'}, 'post');
```

`user33.posts` correctly has the reference to post22. However, `post22.author` is undefined, when I would expect it to reference the `user33` user. Is it possible for datax to do this?

Conversely, I'd want to be able to do the reverse as well. i.e.:

```
const post11  = collection.add({id: 11, title: 'my_tle', author: 33}, 'post');
const post22  = collection.add({id: 22, title: 'my_title', author: 33}, 'post');
const user33  = collection.add({id: 33, email: 'a@b.com'}, 'user');
```
I'd want `user33.posts` to have `post11` and `post22`. However, currently `user33.posts` is an empty list.

Is this even possible? Thanks in advance! Also just want to drop that I'm loving this library. It's the closest thing to ember-data that I've found in non-ember land and it's fairly straightforward to use.

Also using datx version 1.0.2.

**Comment by cooperjbrandon** [comment]
Follow up to my question. On the `User` model, when I change the `posts` property to this:
```
@prop.toMany('post', 'author') public posts!: Post;
```
(i.e add the indirect reference), and then do:
```
collection.add({id: 33, email: 'a@b.com', posts:[22]}, 'user');
```
I get the error `Back references are read only`

**Comment by DarkoKukovec** [comment]
Hi @cooperjbrandon,

You're on the right track with the `@prop.toMany('post', 'author') public posts!: Post;` and I think this is what you need based on the first part of your question.

However, the `Back references are read only` is unfortunately something that can't be avoided in this case. The problem with updating (and setting) back references is that it's quite complex to maintain the data integrity.

An example of this would be if an user had 3 posts, but then it's updated and one of the posts is removed. At this point, the post should not have an author. It's possible you would want this, but I felt that is it quite dangerous and it should be changed directly on the post.
Another (bigger) issue I encountered with this is that the reference order might be important (this is mostly for many to many references) and by modifying the back reference there is no way to control that order. An example here if a post could have multiple authors, it might be important in which order they're listed.

Not sure how (or if?) ember-data is handling this use case (I just recently found out about the similarity, the initial version of this lib was actually Backbone.js influenced), but at some point during development I concluded that allowing this assignment would lead to some logic that wouldn't be consistent and/or intuitive. Perhaps I would need to revisit this topic?

**Comment by cooperjbrandon** [comment]
Ahh interesting. I see what you mean regarding data-integrity and how it's hard to maintain. Although ember-data def handles that. And I've played with [js-data](https://www.js-data.io/) too, and that handles that situation too. Although js-data doesn't have great support with react, and isn't updated a ton, and the modeling is kind of funky, so that's why I've shied away from it.

However, for the issue of reference order on a M2M, I'm fairly certain that's a non issue. Generally what' I've seen is that if there is a required order, there would just be an `order` column on the join-table. And if that's the case, then the FE would just query the join table as well (and then would be represented as an intermediary model in datax or ember-data etc.)

If it's just simple ordering like ABC order, then the developer can just do a computed property with `sort` or something similar (on the FE).

Also, not sure if this helps, but I just checked using js-data (using an old app I created), and by removing a `post` from a user, the `author` field of the removed post becomes undefined. Js-data is it's only standalone lib, so it's probably easier to parse then trying to figure out ember-data. I actually found where this logic occurs, if this helps.
[See here](https://github.com/js-data/js-data/blob/17020d89154651aac39b27999d77f951b6090e90/src/DataStore.js#L210) and [here](https://github.com/js-data/js-data/blob/17020d89154651aac39b27999d77f951b6090e90/src/DataStore.js#L268) and [here](https://github.com/js-data/js-data/blob/17020d89154651aac39b27999d77f951b6090e90/src/DataStore.js#L274-L280)
